homework-2/homework2.ts

```diff
@@ -0,0 +1,92 @@
+import express from "express";
+import Joi from "joi";
+import { v4 } from "uuid";
+import { ContainerTypes, ValidatedRequest, ValidatedRequestSchema, createValidator } from "express-joi-validation";
+
+const app = express();
+const validator = createValidator()
+app.use(express.json());
+
+const userSchema = Joi.object({
+    login: Joi.string().required(),
+    password: Joi.string().required().alphanum(),
+    age: Joi.number().required().min(4).max(130),
+});
+
+interface UserSchema extends ValidatedRequestSchema {
+    id: string;
+    login: string;
+    password: string;
+    age: number;
+    isDeleted: boolean;
+}
+
+interface Filter {
+    loginSubstring?: string;
+    limit?:number;
+}
+
+const users: Array<UserSchema> = [];
+
+// CRUD operations
+app.post('/users',
+    validator.query(userSchema),
+    (req: ValidatedRequest<UserSchema>, res) => {
+        const user = req.body as UserSchema
+        user.id = v4();
+        user.isDeleted = false;
+        users.push(user)
+        res.setHeader('Location', `/users/${user.id}`);
+        res.status(200).send(user);
+    })
+
+app.get('/users', (req, res) => {
+    const filter = req.query as Filter
+    filter.loginSubstring = ''
+    filter.limit = 10
+    const list = users
+        .filter(user => user.login.includes(filter.loginSubstring!) && !user.isDeleted)
```

Comment:
When I execute this GET endpoint, got this error, you can use optional chaining to avoid this
![WeChat3c5f46b2f084ecd428b847146cc8f944](https://user-images.githubusercontent.com/17962458/189303413-b09770d5-e797-4b02-a003-5ea8fee79217.png)